hoomd/mpcd/RejectionVirtualParticleFillerGPU.cuh

```diff
@@ -0,0 +1,265 @@
+ * with system temperature. Along with this a boolean array tracks if the particles are in/out of bounds of the
+ * given geometry.
+ */
+template<class Geometry>
+__global__ void draw_virtual_particles(Scalar4 *d_tmp_pos,
+                                       Scalar4 *d_tmp_vel,
+                                       bool *d_track_bounded_particles,
+                                       const Scalar3 lo,
+                                       const Scalar3 hi,
+                                       const unsigned int first_tag,
+                                       const Scalar vel_factor,
+                                       const unsigned int filler_id,
+                                       const unsigned int type,
+                                       const unsigned int N_virt_max,
+                                       const unsigned int timestep,
+                                       const unsigned int seed,
+                                       const unsigned int block_size,
+                                       const Geometry geom)
+    {
+    // one thread per particle
+    const unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx >= N_virt_max)
+        return;
+
+    // initialize random number generator for positions and velocity
+    hoomd::RandomGenerator rng(hoomd::RNGIdentifier::RejectionFiller, seed, timestep, first_tag+idx, filler_id);
+    d_tmp_pos[idx] = make_scalar4(hoomd::UniformDistribution<Scalar>(lo.x, hi.x)(rng),
+                                  hoomd::UniformDistribution<Scalar>(lo.y, hi.y)(rng),
+                                  hoomd::UniformDistribution<Scalar>(lo.z, hi.z)(rng),
+                                  __int_as_scalar(type));
+
+    hoomd::NormalDistribution<Scalar> gen(vel_factor, 0.0);
+    Scalar3 vel;
+    gen(vel.x, vel.y, rng);
+    vel.z = gen(rng);
+    d_tmp_vel[idx] = make_scalar4(vel.x,
+                                  vel.y,
+                                  vel.z,
+                                  __int_as_scalar(mpcd::detail::NO_CELL));
+
+    // check if particle is inside/outside the confining geometry
+    d_track_bounded_particles[idx] = geom.isOutside(make_scalar3(d_tmp_pos[idx].x,
+                                                                 d_tmp_pos[idx].y,
+                                                                 d_tmp_pos[idx].z));
+    }
+
+/*!
+ * \b implementation
+ * Using one thread per particle, we assign the particle position, velocity and tags using the compacted indices
+ * array as an input.
+ */
+__global__ void parallel_copy(unsigned int *d_compact_indices,
+                              Scalar4 *d_permanent_positions,
+                              Scalar4 *d_permanent_velocities,
+                              unsigned int *d_permanent_tags,
+                              Scalar4 *d_temporary_positions,
+                              Scalar4 *d_temporary_velocities,
```

Comment:
These should be pointers to const, it can allow the compiler to make optimizations related to aliasing:
```suggestion
                              const Scalar4 *d_temporary_positions,
                              const Scalar4 *d_temporary_velocities,
```